Sentence structure changes and ISentenceFilter

suflors opened the pull request:
Changes the structure of Sentence and the OIDs inside of it, adds interface IMarkovFilter featuring basic filtering options that will be available in any implementation.

suflors committed:
added IFilterHandler

suflors committed:
Merged Develop change of AT.Markov.Database to AT.Markov.Data

suflors committed:
required ISentenceAccess methods to take an IFilterHandler as a parameter, updated Generate() to take a filter in order to pass it to ISentenceAccess methods

suflors committed:
updated MarkovTests to use a mock filter for DatabaseFailure()

suflors committed:
merged develop onto feature/filter and handled merge conflicts

suflors committed:
Merge branch 'develop' into feature/filter

suflors committed:
Merge branch 'develop' into feature/filter

suflors committed:
Merge branch 'develop' into feature/filter

suflors committed:
Added Service, Guild, Channel, Subchannel, and User to IFilterHandler.

suflors committed:
Renamed IFilterHandler to IMarkovFilter. This filter will only be used for generation of Markov sentences, hence the name.

suflors committed:
removed unecessary User parameter in IMarkovFilter

suflors committed:
Added Instance parameter to IMarkovFilter

suflors committed:
Removed User from ObjectOID and changed Fragment to an int.

suflors committed:
Made ObjectOID.Instance, ObjectOid.Guild, ObjectOID.Channel, ObjectOID.Subchannel, and AuthorOID.Instance.

suflors committed:
Changed Sentence formatting for parity with ObjectOID and AuthorOID

suflors committed:
Made ObjectOID.Message and ObjectOID.Fragment nullable

suflors committed:
Made AuthorOID.Author nullable.

suflors committed:
changed ObjectOID to an interface with less members

suflors committed:
Updated IMarkovFilter to use IObjectOID

suflors committed:
Added BookObjectOID with parameters and constructors but no implemented functions, edited ServiceType RawText to be Book.

suflors committed:
Added ObjectOIDEscapements.

suflors committed:
Added ObjectOIDEscapementTest class with EscapeTest and UnescapeTest. 16/16 passing.

suflors committed:
Implemented BookObjectOID.ToString and BookObjectOIDTest. 5/5 passing.

suflors committed:
Added BookEscapeTest, passing.

suflors committed:
Moved ObjectOIDEscapement out of the ObjectOID folder. The ObjectOID folder will only be used for implementations of ObjectOID from now on.

suflors committed:
Merged develop onto feature/filter and handled merge conflicts.

suflors committed:
Added ObjectOIDEscapement.Split and tests for it. 6/6 passing.

suflors committed:
Implemented BookObjectOID.Parse().

suflors committed:
Added BookSeriesParseTest(), fixed error where a cast to string[] was used instead of ToArray(). 1/1 passing.

suflors committed:
Added tests with other For functions, 5/5 passing.

suflors committed:
Added BookParseUnescapeTest(), 6/6 passing.

suflors committed:
Added tests to assure that ArgumentExceptions are thrown from BookObjectOID.Parse() when necessary. 4/4 passing.

suflors committed:
Added BookObjectOID.ForInstance() and an option for it in Parse, as well as testing for it.

suflors committed:
Added an ArgumentException for if BookObjectOID.Parse() is given only a ServiceType, as currently Instance is required always.

suflors committed:
Added DiscordObjectOID with unimplemented Parse() and ToString() functions.

suflors committed:
Changed DiscordObjectOID.Fragment to DiscordObjectOID.Sentence to be more clear as to what it is. Added Tests for DiscordObjectOID For functions. 6/6 passing.

suflors committed:
Updated the an error message to specify IObjectOID instead of BookObjectOID as it applies to all IObjectOIDs.

suflors committed:
Implemented DiscordObjectOID.Parse().

suflors committed:
Fixed DiscordObjectOIDTest.DiscordSentenceToStringTest(), which was previously still using ForMessage, 1/1 passing.

suflors committed:
Added DiscordObjectOID.Parse() tests for each For function, 5/5 passing.

suflors committed:
Added DiscordObjectOID tests to test for each ArgumentException that should be thrown.

suflors committed:
Updated DiscordObjectOID to have Category and Thread. Updated DiscordObjectOIDTest to test for these as well, and to fix the existing tests.

suflors committed:
Added a guard in Parse() for both Discord and Book IObjectOID classes for empty strings being passed in, added tests to make sure they work. 2/2 passing.

suflors committed:
Removed nullability for category and thread in For functions in DiscordObjectOID.

suflors committed:
Made AuthorOID.Instance and AuthorOID.Author not nullable. Changed AuthorOID to a class. Add AuthorOID.ToString().

suflors committed:
Changed ObjectOIDEscapement to OIDEscapement.

suflors committed:
Added AuthorOID.Parse().

suflors committed:
Ensured that every member of BookObjectOID, DiscordObjectOID, and AuthorOID are ran through OIDEscapement.Escape() even if it should not be possible for the member to need to be escaped.

suflors committed:
Changed String.Empty to string.Empty in BookParseEmptyTest().

suflors committed:
Added tests for AuthorOID.Parse(). 6/6 passing.

suflors committed:
Removed ID from Sentence, updated all references to ID to use OID or OID.ToString().

suflors committed:
Renamed IMarkovFilter to ISentenceFilter.

suflors committed:
Added IObjectOIDUtil.Parse() and IObjectOIDTest which tests it. 4/4 passing.

suflors committed:
Renamed IObjectOIDTest to IParseableTest. Added IParseable(), OIDParser(), and implementations of IParseable() for each OID ServiceType. 4/4 passing.

suflors committed:
Renamed OIDParser to MultiParser and IParseable to IParser and made MultiParser use <TOID>.

suflors committed:
Separated each ObjectOID Parser into their own classes.

suflors committed:
Updated the error message for failing to parse with MultiParser().

suflors committed:
Made ISentenceFilter into a class SentenceFilter.

suflors committed:
Renamed SentenceFilter.Keyword to SearchString

suflors committed:
Properly overrode AuthorOID.ToString(), replaced where selectors in Discord and Book ObjectOID.ToString()s with OfType<string>(). Implemented a few Intellisense suggestions.

suflors committed:
Changed structure of ToString() functions in Discord and Book ObjectOID.

aliceif committed:
add a customized builder for OIDs and use it

aliceif committed:
implement discord OID parsing using regex

suflors committed:
Edited the error message checks in DiscordParseTooLongTest() and DiscordParseOnlyHasServiceTypeTest() as we no longer give exact error messages when we fail to parse.

suflors committed:
Updated BookObjectOID.ToString() to use OIDBuilder.

suflors committed:
Used the regex method of parsing in BookObjectOID.Parse(). However, this broke escapement being used in parsing, so the regex was edited.

suflors committed:
Removed commented out old Parse() code from BookObjectOID.

suflors committed:
Added OIDPattern, implemented it in BookObjectOID and DiscordObjectOID. Created OIDPatternTest. 2/2 passing.

suflors committed:
Removed leftover handwritten discordOIDRegex from DiscordObjectOID.Parse(). Cleaned up formatting of DiscordObjectOID.Parse().

aliceif commented on AtelierTomato.Markov/AtelierTomato.Markov.Data/Model/OIDPattern.cs:
```diff
@@ -0,0 +1,31 @@
+{
+	public class OIDPattern
+	{
+		public static Regex Generate(IEnumerable<string> fields)
```
If you need indexes, take a type that has indexes (i think that is IList or ICollection), that way you don't need to use ElementAt which is a method that should ideally never be used

aliceif committed:
rework generics in parser types

The concrete parsers should be typed to what they return
In order to be able to use derived types together, the out keyword can be used
TOidCategory is clearer to mean that the multiparser is supposed to take something like IAuthorOid or IObjectOid

aliceif commented on AtelierTomato.Markov/AtelierTomato.Markov.Data.Test/MultiParserTest.cs:
```diff
@@ -0,0 +1,40 @@
+		{
+			MultiParser<IObjectOID> oidParser = new([new BookObjectOIDParser(), new InvalidObjectOIDParser(), new DiscordObjectOIDParser()]);
+			IObjectOID OID = oidParser.Parse("Discord:discord.com:1253189664655806606:1253189664655806610");
+			OID.Should().BeEquivalentTo(DiscordObjectOID.Parse("Discord:discord.com:1253189664655806606:1253189664655806610"));
```
I would also check for expected return type here (that is, DiscordObjectOid), same in equivalent tests

aliceif commented on AtelierTomato.Markov/AtelierTomato.Markov.Generation/GenerateMarkovSentence.cs:
```diff
@@ -4,31 +4,21 @@
 
 			// Tracks the IDs of previously used sentences so that we don't recreate an existing sentence or ping pong between two sentences.
-			var prevIDs = new List<Guid> { sentence.ID };
+			var prevIDs = new List<string> { sentence.OID.ToString() };
```
Can't we make this List\<IObjectOID\>?

suflors replied:
Will implement in SentenceAccess branch, not a change for this pull request.

aliceif reviewed (commented):
Mildly unsure about those empty parser classes, I feel like we could maybe find a way to live without them - or make them be something real and not just wrappers

suflors committed:
Changed OIDPattern to use an IList as a parameter rather than an IEnumerable. closes #5

suflors committed:
Made Discord and Book ObjectOIDTest() check if the OID is of the correct type before checking if the value is correct. closes #6

suflors committed:
Changed all tests that were using Assert() for errors to use Actions. Also updated related errors to report the parameter that caused the error.

aliceif reviewed (approved):
should be ok
